spec.md

```diff
@@ -103,133 +84,162 @@ occurrence that tools and developers can rely upon to better handle the event.
+  * OPTIONAL
 
 ### resource-type
 * Type: String
 * Description: Type of the event source. Providers define list of event sources.
 * Constraints:
-  * Required
+  * REQUIRED
+  * If present, MUST be a non-empty string
 * Examples:
   * s3
 
 ### resource-id
 * Type: String
 * Description: ID of the event source.
 * Constraints:
-  * Required
-*Examples:
-  *my.s3.bucket
+  * REQUIRED
+  * If present, MUST be a non-empty string
+* Examples:
+  * my.s3.bucket
 
 ### event-id
 * Type: String
-* Description: ID of the event.  Can be specified by the producer.  The
-  semantics of this string are explicitly undefined to ease the implementation
-  of producers.  Enables deduplication.
+* Description: ID of the event. The semantics of this string are explicitly
+  undefined to ease the implementation of producers. Enables deduplication.
 * Examples:
   * A database commit ID
 * Constraints:
-  * Required
-  * Unique per producer
+  * REQUIRED
+  * MUST be a non-empty string
+  * MUST be unique within the scope of the producer
 
 ### event-time
-* Type: RFC 3339, timezone Z
+* Type: Timestamp per [RFC 3339](https://tools.ietf.org/html/rfc3339)
 * Description: Timestamp of when the event happened.
+* Constraints:
+  * OPTIONAL
+  * If present, MUST adhere to the format specified in
+    [RFC 3339](https://tools.ietf.org/html/rfc3339)
 
 ### schema-url
-* Type: String
+* Type: URI per [RFC 3986](https://tools.ietf.org/html/rfc3986)
 * Description: A link to the schema which can be optionally specified by the
   producer.
+* Constraints:
+  * OPTIONAL
+  * If present, MUST adhere to the format specified in
+    [RFC 3986](https://tools.ietf.org/html/rfc3986)
 * Notes:
   * It is up for discussion whether this is applicable to the whole event or
     to the data alone.
 
 ### extensions
 * Type: Map <String, Object>
 * Description: This is for additional metadata and this does not have a
-  required structure.  This enables a place for custom fields a producer or
-  middleware may want to include and provides a place to test metadata before
-  adding them to the CloudEvents specification.  TBD - Determine a shorter
+  mandated structure. This enables a place for custom fields a producer or
+  middleware might want to include and provides a place to test metadata before
+  adding them to the CloudEvents specification. TBD - Determine a shorter
   prefix for this (e.g. OpenAPI uses “x-”)
+* Constraints:
+  * OPTIONAL
+  * If present, MUST contain at least one entry
 * Examples:
   * authorization data
   * content type
 
 ### data
 * Type: Arbitrary payload
-* Description: The event payload.  The payload depends on the event-type,
+* Description: The event payload. The payload depends on the event-type,
   schema-url and event-type-version.
+* Constraints:
+  * OPTIONAL
 * Notes:
   * TBD where the producer specifies format/encoding of the data elsewhere in
     this schema.
+  * TBD is it allowed to be empty if present?
 
 ## Context Attributes Backlog
```

Comment:
Note: in today's meeting we decided to move these into bugs, so you might want to remove from this PR to avoid merge conflict.

Reply:
yep - done!